File: q2-backend/ts_template/roundupper_100.ts
import express from 'express';

// location is the simple (x, y) coordinates of an entity within the system
// spaceCowboy models a cowboy in our super amazing system
// spaceAnimal models a single animal in our amazing system
type location = { x: number, y: number };
type spaceCowboy = { name: string, lassoLength: number };
type spaceAnimal = { type: "pig" | "cow" | "flying_burger" };

// spaceEntity models an entity in the super amazing (ROUND UPPER 100) system
type spaceEntity =
    | { type: "space_cowboy", metadata: spaceCowboy, location: location }
    | { type: "space_animal", metadata: spaceAnimal, location: location };


// === ADD YOUR CODE BELOW :D ===

// === ExpressJS setup + Server setup ===
const spaceDatabase = [] as spaceEntity[];
const app = express();
app.use(express.json());

// the POST /entity endpoint adds an entity to your global space database
app.post('/entity', (req, res) => {
    const entities = req.body;
    spaceDatabase.push(...entities.entities);
    res.status(200);
    return res.send("Succesfully added entities to spaceDatabase!");
});

// Helper function to determine if an animal is lassoable.
function is_lassoable(cowboy: spaceEntity, animal: spaceEntity): boolean {
    let cowboy_data = cowboy.metadata as spaceCowboy;
    const lasso_length = cowboy_data.lassoLength;
    const c_l = cowboy.location;
    const a_l = animal.location;
    return lasso_length >= Math.sqrt(Math.pow((c_l.x - a_l.x), 2) + Math.pow((c_l.y - a_l.y), 2));
}

// lassoable returns all the space animals a space cowboy can lasso given their name
app.get('/lassoable', (req, res) => {
    // TODO: fill me in
    const name = req.body.cowboy_name;
    const cowboy: spaceEntity = spaceDatabase.find(entity => entity.type === "space_cowboy" && entity.metadata.name === name) as spaceEntity;

    if (cowboy === undefined) {
        res.status(400);
        return res.send(`No cowboy matching the given name ${name}`);
    }

    const animals: any[] = [];
    for (const e of spaceDatabase) {
        if (e.type === "space_animal" && is_lassoable(cowboy, e) === true) {
            animals.push({
                "type": e.metadata.type,
                "location": e.location
            });
        }
    }

    const result = { "space_animals": animals };
    console.log(result);
    res.status(200);
    return res.send("Bruh");
})

app.listen(8080);

File: q2-backend/py_template/roundupper_100.py
from dataclasses import dataclass
from enum import Enum
from typing import Union, NamedTuple, List
from flask import Flask, request, Response
from math import sqrt
from json import dumps
URL = 'localhost:8080'

# SpaceCowboy models a cowboy in our super amazing system
@dataclass
class SpaceCowboy:
    name: str
    lassoLength: int

# SpaceAnimal models a single animal in our amazing system
@dataclass
class SpaceAnimal:
    # SpaceAnimalType is an enum of all possible space animals we may encounter
    class SpaceAnimalType(Enum):
        PIG = "pig"
        COW = "cow"
        FLYING_BURGER = "flying_burger"

    type: SpaceAnimalType

# SpaceEntity models an entity in the super amazing (ROUND UPPER 100) system
@dataclass
class SpaceEntity:
    class Location(NamedTuple):
        x: int
        y: int

    metadata: Union[SpaceCowboy, SpaceAnimal]
    location: Location
    
# === Helper Function ===

def in_range(cowboy, animal):
    lasso_length = cowboy.metadata.lassoLength
    c_x, c_y = int(cowboy.location['x']), int(cowboy.location['y'])
    a_x, a_y = int(animal.location['x']), int(animal.location['y'])
    return lasso_length >= sqrt((c_x - a_x)**2 + (c_y - a_y)**2)

# ==== HTTP Endpoint Stubs ====
app = Flask(__name__)
space_database: List[SpaceEntity] = []

# the POST /entity endpoint adds an entity to your global space_database
@app.route('/entity', methods=['POST'])
def create_entity():
    entities = request.get_json()
    for info in entities['entities']:
        # Could edit the entity class/create a factory class in case there are more space_entities added later
        type = info['type']
        metadata = info['metadata']
        location = info['location']
        # Create a new SpaceEntity
        if type == 'space_cowboy':
            space_database.append(SpaceEntity(SpaceCowboy(metadata['name'], metadata['lassoLength']), location))
        elif type == 'space_animal':
            space_database.append(SpaceEntity(SpaceAnimal(metadata['type']), location))
        
        
    return Response(status=200)
       

@app.route('/list', methods=['GET'])
def list_entities():
    # For testing purposes...
    return space_database

# lasooable returns all the space animals a space cowboy can lasso given their name
@app.route('/lassoable', methods=['GET'])
def lassoable():
    
    info = request.get_json()
    # Search through the list of SpaceEntites 
    name = info['cowboy_name']
    cowboy = [entity for entity in space_database if type(entity.metadata) == SpaceCowboy and entity.metadata.name == name]
    if len(cowboy) == 0:
        print(f"Failed to find a cowboy with name {name}")
        return Response(status=400)
    
    cowboy = cowboy[0]
    animals = [animal for animal in space_database if type(animal.metadata) == SpaceAnimal and in_range(cowboy, animal)]
    
    
    # Formatting 
    res = []
    for animal in animals:
        res.append({
            "type" : animal.metadata.type,
            "location" : {
                "x" : animal.location['x'],
                "y" : animal.location['y']
            }
        })
    
    print({"space_animals" : res})
    return Response(status=200)
    

# DO NOT TOUCH ME, thanks :D
if __name__ == '__main__':
    app.run(debug=True, port=8080)